File: client/java/Pirate.java
import com.sun.jna.Pointer;

///A basic pirate. These units are bound to land unless aboard a ship. they can pickup and drop treasure as well as build ports and fight other pirates.
class Pirate extends Unit 
{
  public Pirate(Pointer p)
  {
    super(p);
  }
  boolean validify()
  {
    if(iteration == BaseAI.iteration) return true;
    for(int i = 0; i < BaseAI.pirates.length; i++)
    {
      if(BaseAI.pirates[i].ID == ID)
      {
        ptr = BaseAI.pirates[i].ptr;
        iteration = BaseAI.iteration;
        return true;
      }
    }
    throw new ExistentialError();
  }
    
    //commands
    
  ///Move the unit to the designated X and Y coordinates if possible
  int move(int x, int y)
  {
    validify();
    return Client.INSTANCE.pirateMove(ptr, x, y);
  }
  ///Allows a unit to display a message to the screen.
  int talk(String message)
  {
    validify();
    return Client.INSTANCE.pirateTalk(ptr, message);
  }
  ///Attempt to attack the input target if possible
  int attack(Unit Target)
  {
    validify();
    Target.validify();
    return Client.INSTANCE.pirateAttack(ptr, Target.ptr);
  }
  ///Allows the pirate to pickup treasure on the ground.
  int pickupTreasure(int amount)
  {
    validify();
    return Client.INSTANCE.piratePickupTreasure(ptr, amount);
  }
  ///Allows the pirate to drop treasure they are carrying.
  int dropTreasure(int amount)
  {
    validify();
    return Client.INSTANCE.pirateDropTreasure(ptr, amount);
  }
  ///Pirate builds a port on a land tile with water tile adjacent. Cannot be within three spaces of another port!
  int buildPort()
  {
    validify();
    return Client.INSTANCE.pirateBuildPort(ptr);
  }
    
    //getters
    
  ///Unique Identifier
  public int getId()
  {
    validify();
    return Client.INSTANCE.pirateGetId(ptr);
  }
  ///The X position of this object.  X is horizontal, with 0,0 as the top left corner
  public int getX()
  {
    validify();
    return Client.INSTANCE.pirateGetX(ptr);
  }
  ///The Y position of this object.  Y is vertical, with 0,0 as the top left corner
  public int getY()
  {
    validify();
    return Client.INSTANCE.pirateGetY(ptr);
  }
  ///Represents the owner of the unit.
  public int getOwner()
  {
    validify();
    return Client.INSTANCE.pirateGetOwner(ptr);
  }
  ///Current ealth of the unit
  public int getHealth()
  {
    validify();
    return Client.INSTANCE.pirateGetHealth(ptr);
  }
  ///Attacking strength of the unit (Each point of strength deals 1 health of damage)
  public int getStrength()
  {
    validify();
    return Client.INSTANCE.pirateGetStrength(ptr);
  }
  ///Displays the remaining moves for this unit this turn
  public int getMovesLeft()
  {
    validify();
    return Client.INSTANCE.pirateGetMovesLeft(ptr);
  }
  ///Displays the remaining attacks for this unit this turn
  public int getAttacksLeft()
  {
    validify();
    return Client.INSTANCE.pirateGetAttacksLeft(ptr);
  }
  ///Amount of gold carried by the unit.
  public int getGold()
  {
    validify();
    return Client.INSTANCE.pirateGetGold(ptr);
  }

}

File: client/c/Pirate.cpp
// -*-c++-*-

#include "Pirate.h"
#include "game.h"

#include "Unit.h"

Pirate::Pirate(_Pirate* pointer)
{
    ptr = (void*) pointer;
}

int Pirate::id()
{
  return ((_Pirate*)ptr)->id;
}

int Pirate::x()
{
  return ((_Pirate*)ptr)->x;
}

int Pirate::y()
{
  return ((_Pirate*)ptr)->y;
}

int Pirate::owner()
{
  return ((_Pirate*)ptr)->owner;
}

int Pirate::health()
{
  return ((_Pirate*)ptr)->health;
}

int Pirate::strength()
{
  return ((_Pirate*)ptr)->strength;
}

int Pirate::movesLeft()
{
  return ((_Pirate*)ptr)->movesLeft;
}

int Pirate::attacksLeft()
{
  return ((_Pirate*)ptr)->attacksLeft;
}

int Pirate::gold()
{
  return ((_Pirate*)ptr)->gold;
}


int Pirate::move(int x, int y)
{
  return pirateMove( (_Pirate*)ptr, x, y);
}

int Pirate::talk(char* message)
{
  return pirateTalk( (_Pirate*)ptr, message);
}

int Pirate::attack(Unit& Target)
{
  return pirateAttack( (_Pirate*)ptr, (_Unit*) Target.ptr);
}

int Pirate::pickupTreasure(int amount)
{
  return piratePickupTreasure( (_Pirate*)ptr, amount);
}

int Pirate::dropTreasure(int amount)
{
  return pirateDropTreasure( (_Pirate*)ptr, amount);
}

int Pirate::buildPort()
{
  return pirateBuildPort( (_Pirate*)ptr);
}



std::ostream& operator<<(std::ostream& stream,Pirate ob)
{
  stream << "id: " << ((_Pirate*)ob.ptr)->id  <<'\n';
  stream << "x: " << ((_Pirate*)ob.ptr)->x  <<'\n';
  stream << "y: " << ((_Pirate*)ob.ptr)->y  <<'\n';
  stream << "owner: " << ((_Pirate*)ob.ptr)->owner  <<'\n';
  stream << "health: " << ((_Pirate*)ob.ptr)->health  <<'\n';
  stream << "strength: " << ((_Pirate*)ob.ptr)->strength  <<'\n';
  stream << "movesLeft: " << ((_Pirate*)ob.ptr)->movesLeft  <<'\n';
  stream << "attacksLeft: " << ((_Pirate*)ob.ptr)->attacksLeft  <<'\n';
  stream << "gold: " << ((_Pirate*)ob.ptr)->gold  <<'\n';
  return stream;
}

